src/hotspot/cpu/x86/macroAssembler_x86_aes.cpp

```diff
@@ -1267,4 +1267,627 @@ void MacroAssembler::aesctr_encrypt(Register src_addr, Register dest_addr, Regis
+    // Second phase of reduction
+    if (final_reduction) {
+        evpclmulqdq(ZTMP9, ZTMP12, ZTMP10, 0x00, Assembler::AVX_128bit);
+        vpsrldq(ZTMP9, ZTMP9, 4, Assembler::AVX_128bit); // Shift-R 1-DW to obtain 2-DWs shift-R
+        evpclmulqdq(ZTMP11, ZTMP12, ZTMP10, 0x10, Assembler::AVX_128bit);
+        vpslldq(ZTMP11, ZTMP11, 4, Assembler::AVX_128bit); // Shift-L 1-DW for result
+        // ZTMP5 = ZTMP5 X ZTMP11 X ZTMP9
+        vpternlogq(ZTMP5, 0x96, ZTMP11, ZTMP9, Assembler::AVX_128bit);
+    }
+    // Last AES round
+    lastroundEncode(ZTMP17, ZTMP0, ZTMP1, ZTMP2, ZTMP3);
+    // XOR against plain / cipher text
+    xorBeforeStore(ZTMP0, ZTMP1, ZTMP2, ZTMP3, ZTMP13, ZTMP14, ZTMP15, ZTMP16);
+    // store cipher / plain text
+    storeData(out, pos, ZTMP0, ZTMP1, ZTMP2, ZTMP3);
+}
+
+void MacroAssembler::aesgcm_encrypt(Register in, Register len, Register ct, Register out, Register key,
+                                    Register state, Register subkeyHtbl, Register counter) {
+
+    Label ENC_DEC_DONE, GENERATE_HTBL_48_BLKS, AES_192, AES_256, STORE_CT, GHASH_LAST_32,
+          AES_32_BLOCKS, GHASH_AES_PARALLEL, LOOP, ACCUMULATE, GHASH_16_AES_16;
+    const XMMRegister CTR_BLOCKx = xmm9;
+    const XMMRegister AAD_HASHx = xmm14;
+    const Register pos = rax;
+    const Register rounds = r15;
+    Register ghash_pos;
+#ifndef _WIN64
+    ghash_pos = r14;
+#else
+    ghash_pos = r11;
+#endif // !_WIN64
+    const XMMRegister ZTMP0 = xmm0;
+    const XMMRegister ZTMP1 = xmm3;
+    const XMMRegister ZTMP2 = xmm4;
+    const XMMRegister ZTMP3 = xmm5;
+    const XMMRegister ZTMP4 = xmm6;
+    const XMMRegister ZTMP5 = xmm7;
+    const XMMRegister ZTMP6 = xmm10;
+    const XMMRegister ZTMP7 = xmm11;
+    const XMMRegister ZTMP8 = xmm12;
+    const XMMRegister ZTMP9 = xmm13;
+    const XMMRegister ZTMP10 = xmm15;
+    const XMMRegister ZTMP11 = xmm16;
+    const XMMRegister ZTMP12 = xmm17;
+    const XMMRegister ZTMP13 = xmm19;
+    const XMMRegister ZTMP14 = xmm20;
+    const XMMRegister ZTMP15 = xmm21;
+    const XMMRegister ZTMP16 = xmm30;
+    const XMMRegister COUNTER_INC_MASK = xmm18;
+
+    movl(pos, 0); // Total length processed
+    // Min data size processed = 768 bytes
+    cmpl(len, 768);
+    jcc(Assembler::less, ENC_DEC_DONE);
+
+    // Generate 48 constants for htbl
+    call(GENERATE_HTBL_48_BLKS, relocInfo::none);
+    int index = 0; // Index for choosing subkeyHtbl entry
+    movl(ghash_pos, 0); // Pointer for ghash read and store operations
+
+    // Move initial counter value and STATE value into variables
+    movdqu(CTR_BLOCKx, Address(counter, 0));
+    movdqu(AAD_HASHx, Address(state, 0));
+    // Load lswap mask for ghash
+    movdqu(xmm24, ExternalAddress(StubRoutines::x86::ghash_long_swap_mask_addr()), rbx);
+    // Shuffle input state using lswap mask
+    vpshufb(AAD_HASHx, AAD_HASHx, xmm24, Assembler::AVX_128bit);
+
+    // Compute #rounds for AES based on the length of the key array
```

Comment:
This is a bit of a hack. Wouldn't it make more sense to pass in the array oop, then derive both the length and the address of the base of the key array from the oop, rather than using a negative offset from the base address?

Reply:
Forget that, it's done everywhere. Deleted.